sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -293,20 +294,118 @@ case class Substring(str: Expression, pos: Expression, len: Expression)
 }
 
 /**
- * A function that return the length of the given string expression.
+ * A function that return the length of the given string or binary expression.
  */
-case class StringLength(child: Expression) extends UnaryExpression with ExpectsInputTypes {
+case class Length(child: Expression) extends UnaryExpression with ExpressionConstraint {
   override def dataType: DataType = IntegerType
-  override def expectedChildTypes: Seq[DataType] = Seq(StringType)
+  def constraint: Seq[DataTypeConstraint] =
+    DataTypeConstraint(AcceptSpecifiedType(Set(StringType, BinaryType))) :: Nil
```

Comment:
It's supposed not to be done in `Length`, see https://issues.apache.org/jira/browse/SPARK-8230